cpp/src/arrow/util/io_util.cc

```diff
@@ -1115,7 +1116,9 @@ Result<FileDescriptor> FileOpenWritable(const PlatformFilename& file_name,
   fd = FileDescriptor(ret);
 #else
   int oflag = O_CREAT;
-
+  if (sync) {
+    oflag |= O_SYNC;
```

Comment:
Just for discussion.
My understanding is O_SYNC (O_DSYNC might be better) is necessary as fadvise won't drop dirty pages.
If that's the only reason to add this flag, looks to me it's better sync data explicitly before fadvise.
```
write(fd, buffer, size);
fdatasync(fd);   // explicitly flush fd
fadvise(fd, dontneed);
```

Reply:
Yes I think O_DSYNC probably makes more sense. I think there might be other cases in which you want to open a file with O_DSYNC where you are not using fadvise. So perhaps having a separate flag here is good for future use cases. Although I am struggling to think of one right now.

Reply:
Can you explain why fdatasync would be better? Seems like it's additional system call.

Reply:
It's purely from coding perspective.
As `O_SYNC` is rarely used, avoiding the flag and adding `fdatasync` explicitly before `fadvice` make the code easier to read IMHO.
An additional syscall is probably not a worry. I think the use case is infrequent flushing of large data block.